File: src/lib/canvas/editor-session/dirty-notifier.ts
import type { AutoSave } from '$lib/session/auto-save';

/**
 * Port that editor-session layers (TabState, Workspace) use to signal that
 * persistable state has changed. Implementations decide how to react — e.g.
 * the production adapter debounces through `AutoSave`, while tests use a
 * recording fake.
 *
 * Production adapter: `createAutoSaveDirtyNotifier(autoSave)`.
 * Test adapter: `createFakeDirtyNotifier()` from `./fake-dirty-notifier.ts`.
 */
export interface DirtyNotifier {
	/** A document's persistable state changed. */
	markDirty(documentId: string): void;
	/** A document was removed; flush any pending state for it. */
	notifyTabRemoved(documentId: string): void;
}

/**
 * Wraps an `AutoSave` instance as a `DirtyNotifier`. The notifier owns no
 * state of its own — every call delegates to the enclosed `AutoSave`.
 */
export function createAutoSaveDirtyNotifier(autoSave: AutoSave): DirtyNotifier {
	return {
		markDirty: (documentId) => autoSave.markDirty(documentId),
		notifyTabRemoved: (documentId) => autoSave.notifyTabRemoved(documentId)
	};
}

File: apple/Dotorixel/Persistence/DirtyNotifier.swift
/// Port that the state layers (`Workspace`, `TabState`, `SharedState`) use
/// to signal that persistable state changed (web parity: `DirtyNotifier` in
/// `dirty-notifier.ts`). Implementations decide how to react — production
/// debounces through `AutoSave`; tests record.
protocol DirtyNotifier {
    /// A document's persistable state changed.
    func markDirty(documentId: String)
    /// A document's tab closed; its record must not be rewritten.
    func notifyTabRemoved(documentId: String)
}

/// The default notifier for workspaces opened without persistence wiring
/// (previews, tests): every signal is dropped.
struct NoOpDirtyNotifier: DirtyNotifier {
    func markDirty(documentId: String) {}
    func notifyTabRemoved(documentId: String) {}
}
